### PR TITLE
daemon: use 'private' ipc mode by default

### PR DESCRIPTION
**Pull request by kolyshkin**
_(this is a follow-up to https://github.com/moby/moby/pull/34087)_

This changes the default ipc mode of daemon/engine to be private, meaning the containers will not have their `/dev/shm` filesystem mounted on the host and then bind-mounted from host to a container by default.

The benefits of doing this are:

1. **No leaked mounts.** This eliminates a possibility to leak mounts into other namespaces (and therefore unfortunate errors like "Unable to remove filesystem for $ID: remove /var/lib/docker/containers/$ID/shm: device or resource busy").

2. **Working checkpoint/restore.** Make `docker checkpoint` not lose the contents of `/dev/shm`, but save it to the dump, and be restored back upon `docker start --checkpoint`.

3. **Better security.** Currently any container is opened to share its `/dev/shm` with any other container.

The downside is, obviously, this change will break the following usage scenario:
```
 $ docker run -d --name donor busybox top
 $ docker run --rm -it --ipc container:donor busybox sh
 Error response from daemon: linux spec namespaces: can't join IPC
 of  container <ID>: non-shareable IPC (hint: use --ipc shareable
 for the donor container)
```
The solution, as hinted by the (amended) error message, is to explicitly enable donor sharing by using `--ipc shareable` option to `docker run` or `docker create`:
```
 $ docker run -d --name donor --ipc shareable busybox top
```

## Compatibility notes

1. This only applies to containers created _after_ this change. Existing containers are not affected and will work fine as their ipc mode is stored in `HostConfig`.

2. Old (bad, but backward-compatible) behavior (i.e. "shareable" containers by default) can be enabled by either using `--default-ipc-mode shareable` daemon command line option, or by adding a `"default-ipc-mode": shareable"` line in `docker.json` configuration file.

3. If an older client (API < 1.40) is used, a "shareable" container is created. A test to check that is added.

4. An item in API version history explaining the new default, and a way to change it back, is added.

5. As a side effect, this PR fixes bug originally addressed in https://github.com/moby/moby/pull/38700 (that is obsoleted by this one).

**- A picture of a cute animal (not mandatory but encouraged)**

## See also 

* First part of this saga: https://github.com/moby/moby/pull/34087
* Documentation updates: https://github.com/docker/cli/pull/367
* Test case fixes that were originally part of this PR: https://github.com/moby/moby/pull/35840
* Kubernetes patch to not rely on IpcMode:shared being the default: https://github.com/kubernetes/kubernetes/pull/70826
* Some more test case updates: https://github.com/moby/moby/pull/38843/commits

screenshot from 2017-11-27 17-05-55
_A  tiger cub from the 1984 USSR cartoon_  https://www.youtube.com/watch?v=0DnjD7w-2FI

### CODE REVIEW

**Comment by kolyshkin** [comment]
Need to figure out why `TestCleanupMountsAfterDaemonAndContainerKill` and `TestCleanupMountsAfterDaemonCrash` fail

**Comment by kolyshkin** [comment]
OK, I got it now why these two tests are failing!

The reason is, all the mounts are now in the private namespace, except for one used for container's /dev/shm. With this patch, this last mount is gone so test is not finding anything.

The solution is to check whether there are any container mounts visible from the host system after container start -- if not, skip the test.

**Comment by kolyshkin** [comment]
This one is now ready for the prime time

**Comment by kolyshkin** [comment]
Powerpc test case failure is unrelated, nevertheless I figured it out and fixed the test case.

**Comment by kolyshkin** [comment]
The following failure is most probably unrelated as I am seeing it in other PRs. I'll take a look nevertheless.

```
12:29:56 ----------------------------------------------------------------------
12:29:56 FAIL: check_test.go:366: DockerSwarmSuite.TearDownTest
12:29:56 
12:29:56 check_test.go:371:
12:29:56     d.Stop(c)
12:29:56 daemon/daemon.go:395:
12:29:56     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
12:29:56 ... Error: Error while stopping the daemon d704ce7bc07d4 : exit status 130
12:29:56 
12:29:56 
12:29:56 ----------------------------------------------------------------------
12:29:56 PANIC: docker_api_swarm_service_test.go:201: DockerSwarmSuite.TestAPISwarmServicesUpdateStartFirst
12:29:56 
12:29:56 [d704ce7bc07d4] waiting for daemon to start
12:29:56 [d704ce7bc07d4] daemon started
12:29:56 
12:29:56 [d704ce7bc07d4] daemon started
12:29:56 Attempt #2: daemon is still running with pid 19775
12:29:56 Attempt #3: daemon is still running with pid 19775
12:29:56 Attempt #4: daemon is still running with pid 19775
12:29:56 [d704ce7bc07d4] exiting daemon
12:29:56 ... Panic: Fixture has panicked (see related PANIC)
12:29:56 
12:29:56 ----------------------------------------------------------------------
```

**Comment by cpuguy83** [comment]
I worry about changing the default.
If the goal is to eliminate issues with EBUSY, shouldn't we dig deeper into the cause of these leaks? This isn't really fixing mount leaks, just making it less likely in the default configuration.

File: daemon/container_operations_unix.go
Comment (thaJeztah) [inline daemon/container_operations_unix.go]: We should avoid mentioning flags here :smile:

**Comment by cpuguy83** [comment]
We discussed this in the maintainers meeting and it came up "well is Kubernetes depending on shared IPC" ... I believe it is and this will break them... and even if not it's a good example of how this could break people.

**note**: I mentioned on the call maybe they are using exec but that just doesn't make any sense because they couldn't start containers with different images in the pod, which is required for the pod scenario.

**Comment by kolyshkin** [comment]
Indeed Kubernetes uses `IpcMode` of `container:$SANDBOX`, where `$SANDBOX` is a container running some "pause" image. Currently it uses API 1.31 so it won't break but once they'll revendor it will.

It can be fixed in k8s code (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/docker_sandbox.go), in particular `func (ds *dockerService) makeSandboxDockerConfig` needs to amend HostConfig with `IpcMode: dockercontainer.IpcMode("shareable")`. I'll see to it.

Because of this, yes, I think we should NOT merge this for now.

**Comment by thaJeztah** [comment]
Thanks @kolyshkin !

Meanwhile, we could add something in the documentation to make "private" a recommendation (if you don't need sharing)

**Comment by ngrilly** [comment]
@thaJeztah Is there an issue opened regarding the errors "Unable to remove filesystem for $ID: remove /var/lib/docker/containers/$ID/shm: device or resource busy"? I was unable to find one still opened.

**Comment by kolyshkin** [comment]
Maybe it's time to revisit this one. I have submitted the patch to Kube here: https://github.com/kubernetes/kubernetes/pull/70826

File: daemon/container_operations_unix.go
Comment (kolyshkin) [reply daemon/container_operations_unix.go]: done (it's "IpcMode: shareable" now)

**Comment by kolyshkin** [comment]
The majority of test changes that were part of this PR were moved to #35840 (merged long time ago).

**Comment by olljanat** [comment]
@kolyshkin FYI CI build looks failing to: 
```
18:38:39     --- FAIL: TestDockerNetworkIpvlan/L2InternalMode (0.96s)
18:38:39         ipvlan_test.go:181: assertion failed: err is nil
18:38:39         ipvlan_test.go:182: assertion failed: expression is false: timeoutCtx.Err() == context.DeadlineExceeded
18:38:39     daemon.go:283: [db9e063d4ebfc] exiting daemon
18:38:39     daemon.go:296: [d961fa4954580] waiting for daemon to start
```
(if this is one still valid after all)

**Comment by olljanat** [comment]
Derek add label: rebuild/*

**Comment by olljanat** [comment]
Derek add label: status/failing-ci

File: api/server/router/container/container_routes.go
Comment (cpuguy83) [inline api/server/router/container/container_routes.go]: Sadly this API version is a bit out of date.

**Review by cpuguy83 — COMMENTED** [review]
Change seems fine once API version is updated and tests come clean.

I see also that the k8s change was merged.

File: api/server/router/container/container_routes.go
Comment (kolyshkin) [reply api/server/router/container/container_routes.go]: > Sadly this API version is a bit out of date.

￼You mean for the case of k8s client? Originally this patch was intended for compatibility with older docker cli which was not able to set "shareable".

It looks like kubernetes is currently using docker api 1.38. The problem is, we can't have separate cases for docker-cli, kubelet client and others.

I can't think of any easy solution here, rather than wait a few years and assume the fix (https://github.com/kubernetes/kubernetes/pull/70826) has made its way to all k8s versions, then change the default to be private.

Comment (cpuguy83) [reply api/server/router/container/container_routes.go]: Shouldn't this just be gated to the current API version?

Comment (kolyshkin) [reply api/server/router/container/container_routes.go]: I don't think so. The meaning of this hunk is to use old default of "shareable" for older clients (i.e. those that have no way to set "shareable" ipcmode explicitly). As the functionality to set shareable ipcmode appeared in API v1.32, the check is for < 1.32.

Comment (cpuguy83) [reply api/server/router/container/container_routes.go]: Sorry, I don't think we should change the default without gating it by the api version where we've changed the default.

Comment (kolyshkin) [reply api/server/router/container/container_routes.go]: OK, so we have a server (daemon, engine) and a client, and an API version.

Since API 1.32
1. the client can explicitly set shareable IPC mode;
2. the server can have default IPC mode set to shareable (using `default-ipc-mode` in either daemon.json, or as a command line option).

The code in question was added as a client with API < 1.32 is unable to set shareable IPC mode at all, not because the default is changed. Strictly speaking, it does not belong to this PR, but is rather a followup to #34087.

As far as I understand, if we change `1.32` in the code above to a later version, the above functionality (2) of setting default IPC mode will cease to work.

Comment (cpuguy83) [reply api/server/router/container/container_routes.go]: My point is, changing the default should be gated by the API version we are introducing the change in rather than some old API version. So basically we shouldn't send an empty value for any version < the current version and it should be the old default ("shareable").

Comment (cpuguy83) [reply api/server/router/container/container_routes.go]: We seem to have a disconnect here:

< 1.32 there was no option and only shared was used
>= 1.32 default is shared but client can change

My proposal is to say if version is >= 1.32 < 1.40 (or whatever the current released API version is at this point...) to keep the default as shareable, and for the upcoming release, daemon default of private.

Comment (thaJeztah) [reply api/server/router/container/container_routes.go]: Yes, I'm on @cpuguy83's side of this; if we change this value for API v1.32..v1.40 in the upcoming release, it means that users that are _currently_ using API v1.32..v1.40 that update the daemon will all of a sudden get a different behavior on the same API version.

Basically, if a user pins to API version x, then updating the daemon to whatever newer version should maintain exactly the same behavior (baring bug-fixes of course). If they do _not_ pin to a specific version (or explicitly update to pin to a newer API version) they are ok with getting new behavior introduced by the newer API version.

Comment (thaJeztah) [reply api/server/router/container/container_routes.go]: Of course, we can recommend somewhere that users should configure their daemon to;

- use `private` or `shareable` as a default if they're using API v1.32 or up if they do not require `shared`
- or (the reverse); configure their daemon to use `shared` as default before updating to the upcoming release that has this change (if they want to use API v1.40+ but need to keep the old behavior) (this one should definitely go into the release notes)

Comment (kolyshkin) [reply api/server/router/container/container_routes.go]: Perhaps I made a very wrong decision to put these two commits together in one PR, and this is the source of confusion. Let me remove the second commit and put it into a separate PR, hoping it will be easier to grasp that way.

**Comment by kolyshkin** [comment]
I have split the second commit from this PR into #38700, as it does not really belong here, and for the sake of easier review.

File: api/server/router/container/container_routes.go
Comment (kolyshkin) [reply api/server/router/container/container_routes.go]: #38700

**Comment by cpuguy83** [comment]
@kolyshkin This wasn't confusing at all. Just that it's wrong to break existing users, which is what this does.

**Comment by kolyshkin** [comment]
We disscused this with @thaJeztah and @cpuguy83 on maintainers meeting today. What needs to be added is:
* a changelog entry explaining the new default;
* a fix for clients API < 1.41: if new container ipc mode is empty AND compiled-in daemon default is not changed, use "shareable".

**Comment by thaJeztah** [comment]
For the API version; upcoming release will be API v1.40, so if it gets merged before that, it's gonna be `API < 1.40`

**Comment by kolyshkin** [comment]
I'll revisit this one once #38700 is merged.

**Comment by kolyshkin** [comment]
OK I just came to realization that #38700 needs to be merged into this one, for the reasons that I am too lazy to describe right now. Also that @cpuguy83 and @thaJeztah was right from the start asking to forcefully set `IpcMode: shareable` for older clients.

Currently this includes #38843; I will rebase once it's merged.

**Comment by kolyshkin** [comment]
Derek add label: rebuild/janky

> 01:16:03.313 FAIL: docker_api_swarm_test.go:297: DockerSwarmSuite.TestAPISwarmLeaderElection
is a flaky test, rebuilding

**Comment by kolyshkin** [comment]
Derek add label: rebuild/janky

**Comment by kolyshkin** [comment]
Rebased on top of merged #38843

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**
